packages/react/src/components/createInlineOverlayComponent.tsx

```diff
@@ -121,26 +145,48 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
         ...propsToPass,
         ref: this.stableMergedRefs,
         style,
+        key: trackByKey,
       };
 
-      /**
-       * We only want the inner component
-       * to be mounted if the overlay is open,
-       * so conditionally render the component
-       * based on the isOpen state.
-       */
-      return createElement(tagName, newProps, (this.state.isOpen || this.props.keepContentsMounted) ?
-        createElement('div', {
-          id: 'ion-react-wrapper',
-          ref: this.wrapperRef,
-          style: {
-            display: 'flex',
-            flexDirection: 'column',
-            height: '100%'
-          }
-        }, children) :
-        null
-      );
+      return [
+        /**
+         * React will unmount the overlay component when conditional content is
+         * rendered before or after the overlay in the DOM. To work around this
+         * we create a buffer element before and after the overlay component,
+         * so that even if React unmounts those elements, the overlay will still
+         * be in the correct position in the DOM.
+         */
+        createElement('template', { key: `overlay-start-${trackByKey}` }),
+        createElement(
+          tagName,
+          newProps,
+          /**
+           * We only want the inner component
+           * to be mounted if the overlay is open,
+           * so conditionally render the component
+           * based on the isOpen state.
+           */
+          this.state.isOpen || this.props.keepContentsMounted
+            ? createElement(
+                'div',
+                {
+                  id: 'ion-react-wrapper',
+                  ref: this.wrapperRef,
+                  style: {
+                    display: 'flex',
+                    flexDirection: 'column',
+                    height: '100%',
+                  },
+                },
+                children
+              )
+            : null
+        ),
+        createElement('template', {
```

Comment:
The templates before/after the overlay have the additional responsibility of safe guarding against React unmounting the overlay when dynamic nodes are added to or removed.

e.g.:
```tsx
render() {
  const [items, setItems] = useState([]);
  return (
    <>
      {items && items.map(item => `<div>${item}</div>`)}
      <IonModal />
      {items && items.map(item => `<div>${item}</div>`)}
    </>
  );
}
```

In this example, if `items` is mutated, without the buffer elements, React unmounts `<IonModal>`. 

I can explore if we provide our own delegate, if somehow that enables re-inserting the overlay in the DOM before React performs a re-render. This would also address the core delegate currently creating the "ionic teleport" comment and re-inserting the overlay.

Reply:
There isn't an `Node.insertAfter` API, so I don't think React is ever going to reference the "after" template when trying to append new elements. I tried removing it locally and the "IonModalConditionalSibling" test seemed to work.